src/services/mapCode.ts

```diff
@@ -0,0 +1,484 @@
+    originalFile: SourceFile,
+    changeTracker: ChangeTracker,
+    changes: NodeArray<Statement>,
+    focusLocations?: TextSpan[],
+) {
+    // If there's an empty or null focusLocation, we just shove everything at
+    // the end of the file and call it a day.
+    if (!focusLocations?.length) {
+        changeTracker.insertNodesAtEndOfFile(
+            originalFile,
+            changes,
+            /*blankLineBetween*/ false,
+        );
+        return;
+    }
+
+    // Otherwise, we'll need to find the right place to insert or replace.
+
+    // First, we try and find a nearby frame of reference: are there
+    // declarations with similar names nearby?
+    for (const location of focusLocations) {
+        const scope: Node | undefined = findAncestor(
+            getTokenAtPosition(originalFile, location.start),
+            block =>
+                (isBlock(block) || isSourceFile(block)) &&
+                block.statements.some(origStmt => changes.some(newStmt => matchNode(newStmt, origStmt))),
+        );
+        if (scope) {
+            // We found a scope that contains a matching node.
+
+            // First we find the last match of the range we're replacing.
+            // That is, assuming we have the following code to start with:
+            //
+            // ```ts
+            // function foo() {
+            //     console.log("hello");
+            // }
+            // function bar() {
+            //     console.log("world");
+            // }
+            // function baz() {
+            //     console.log("!");
+            // }
+            // function other() {
+            //    console.log("Leave me alone!");
+            // }
+            // ```
+            //
+            // And we get a map request with the following code:
+            //
+            // ```ts
+            // function bar() {
+            //     console.log("world");
+            // }
+            // function baz() {
+            //     console.log("!");
+            // }
+            // function qux() {
+            //     console.log("?");
+            // }
+            // ```
+            //
+            // `bar` will be the beginning of our range, and we want to
+            // have `baz` in the original be the end of our replacement
+            // range. `other` should be left intact, while `qux` should be
+            // "injected" before it.
+            //
+            // We do this by scanning the scope for all nodes that have a
+            // match in the incoming map request code, then sorting and
+            // picking first and last by position, and that will be our
+            // range for replacement.
+            const start = (scope as Block).statements.find(stmt => changes.some(node => matchNode(node, stmt)));
+            const end = findLast((scope as Block).statements, stmt => changes.some(node => matchNode(node, stmt)));
+            // This should generally never be the case at this point, but just in case...

```

Comment:
what should never be the case? `!start || !end` ?

Reply:
correct. I'll clarify the comment.